internal/mvcc/backend/stale_concurrent_read_tx.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	bolt "github.com/coreos/bbolt"
+)
+
+type staleConcurrentReadTx struct {
+	tx *bolt.Tx
+}
+
+func (rt *staleConcurrentReadTx) Lock() {}
```

Comment:
How does `staleConcurrentReadTx` deal with compaction process? IIUC, `Lock()` does nothing so it cannot indicate its existence. What will happen if the compaction process removes keys during the `staleConcurrentReadTx` is reading the keys?

Reply:
@mitake boltdb provides isolation on read/write. basically the read tx gets snapshot of the state. it wont see the writes/deletes happened after it. we need to be careful around the in memory index though. if we discover any key is compacted, we can return ErrCompacted at least so that client can retry.